File: network/src/main/kotlin/org/bitlap/network/processor/BitlapRpcProcessor.kt
package org.bitlap.network.processor

import com.alipay.sofa.jraft.rpc.RpcContext
import com.alipay.sofa.jraft.rpc.RpcRequestClosure
import com.alipay.sofa.jraft.rpc.RpcRequestProcessor
import com.google.protobuf.Message
import org.bitlap.common.logger
import java.util.concurrent.Executor

/**
 * Desc: common abstract for rpc processors
 *
 * Mail: chk19940609@gmail.com
 * Created by IceMimosa
 * Date: 2021/9/4
 */
abstract class BitlapRpcProcessor<T : Message>(
    executor: Executor?,
    private val defaultResp: Message,
) : RpcRequestProcessor<T>(executor, defaultResp), ProcessorHelper {

    val log = logger { }

    override fun handleRequest(rpcCtx: RpcContext, request: T) {
        try {
            val msg = processRequest(request, RpcRequestClosure(rpcCtx, this.defaultResp))
            if (msg != null) {
                rpcCtx.sendResponse(msg)
            }
        } catch (e: Exception) {
            log.error("handleRequest $request failed", e)
            rpcCtx.sendResponse(processError(rpcCtx, e))
        }
    }

    abstract fun processError(rpcCtx: RpcContext, exception: Exception): Message
}

File: network/src/main/scala/org/bitlap/net/processor/BitlapRpcProcessor.scala
package org.bitlap.net.processor

import com.alipay.sofa.jraft.rpc.{ RpcContext, RpcRequestClosure, RpcRequestProcessor }
import com.google.protobuf.Message
import com.typesafe.scalalogging.LazyLogging

import java.util.concurrent.Executor

/**
 *
 * @author 梦境迷离
 * @since 2021/11/21
 * @version 1.0
 */
abstract class BitlapRpcProcessor[T <: Message](executor: Executor, override val defaultResp: Message)
  extends RpcRequestProcessor[T](executor, defaultResp)
    with ProcessorHelper with LazyLogging {

  override def handleRequest(rpcCtx: RpcContext, request: T) {
    try {
      val msg = processRequest(request, new RpcRequestClosure(rpcCtx, this.defaultResp))
      if (msg != null) {
        rpcCtx.sendResponse(msg)
      }
    } catch {
      case e: Exception =>
        logger.error(s"handleRequest $request failed", e)
        rpcCtx.sendResponse(processError(rpcCtx, e))
    }
  }

  def processError(rpcCtx: RpcContext, exception: Exception): Message
}
